chain/ethereum/ethereum.go

```diff
@@ -183,6 +186,66 @@ func ChunkedFilterLogs(ctx context.Context, client ETHClient, addresses []common
 	return failedQueries, nil
 }
 
+// BatchedFilterLogs queries the blockchain for past events in batches.
+// It slices addresses and total number of blocks with pre-defined batch size.
+// The results are later fed into a log chan that was provided by the caller.
+// Failed query intervals are retried after a backoff period, increased exponentially.
+func BatchedFilterLogs(ctx context.Context, client ETHClient, addresses []common.Address, fromBlock, toBlock uint64, logChan chan<- types.Log, backoff time.Duration) {
+	if backoff == 0 || backoff > maxBackoffDuration {
+		backoff = initialBackoffDuration
+	}
+
+	//	Slice and dice whole block interval and smart contract address count so that they fit into predefined maximums.
+	if len(addresses) > addressChunkSize {
+		BatchedFilterLogs(ctx, client, addresses[:addressChunkSize], fromBlock, toBlock, logChan, backoff)
+		BatchedFilterLogs(ctx, client, addresses[addressChunkSize:], fromBlock, toBlock, logChan, backoff)
+		return
+	}
+
+	if toBlock-fromBlock > blockChunkSize {
+		BatchedFilterLogs(ctx, client, addresses, toBlock-blockChunkSize, toBlock, logChan, backoff)
+		BatchedFilterLogs(ctx, client, addresses, fromBlock, toBlock-blockChunkSize-1, logChan, backoff)
+		return
+	}
+
+	log.Debug().Uint64("from", fromBlock).Uint64("to", toBlock).Msg("retrieving historical events...")
+
+	query := ethereum.FilterQuery{
+		FromBlock: big.NewInt(int64(fromBlock)),
+		ToBlock:   big.NewInt(int64(toBlock)),
+		Addresses: addresses,
+	}
+
+	logs, err := client.FilterLogs(ctx, query)
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(backoff):
+			log.Warn().Err(err).Uint64("from", fromBlock).Uint64("to", toBlock).Str("backoff", backoff.String()).Msg("retrying interval")
+
+			// Node providers may refuse the request if they deem it too large.
+			// It could be block range, number of events or just response size.
+			// Here, the call is divided into two, so that it can go through.
+			// If the problem is with some specific block range,
+			// this will also process the good blocks in a binary search fashion.
+			mid := (query.FromBlock.Uint64() + query.ToBlock.Uint64()) / 2
+			BatchedFilterLogs(ctx, client, addresses, fromBlock, mid, logChan, backoff<<1)
+			BatchedFilterLogs(ctx, client, addresses, mid, toBlock, logChan, backoff<<1)
+			return
+		}
+	}
+
+	//	Reset backoff after a successful call
+	if backoff > initialBackoffDuration {
+		backoff = initialBackoffDuration
```

Comment:
I think this is not necessary since backoff will not be used anymore after this line.